### DISCUSSION TITLE
Booting Citrix Hypervisor (XenServer) 8.2 LTSR Express from iPXE

### DISCUSSION THREAD

**Discussion by caeles87** [ipxe/ipxe#670] [Q&A]
Hello,

I am trying to get the [Citrix Hypervisor (XenServer) 8.2 LTSR Express](https://www.citrix.com/downloads/citrix-hypervisor/product-software/hypervisor-82-express-edition.html) installer to be bootable via iPXE. 

I've been trying to follow the [instructions on the iPXE site](https://ipxe.org/appnote/xenserver), but they appear to be from 2014 targeting older versions. In this thread [here](https://gist.github.com/robinsmidsrod/c1d4df4b2e9425d060ec) it is reported that this process stopped working in version 7.2. 

So far what I have is this from that original guide:
```
:citrix-xen-server-8.2-express
imgfree
set xenkernel http://${server_ip}/[...]/CitrixHypervisor-8.2.0-install-cd/boot/xen
set xenvmlinuz http://${server_ip}/[...]/CitrixHypervisor-8.2.0-install-cd/boot/vmlinuz

kernel ${xenkernel} dom0_max_vcpus=2  dom0_mem=1024M,max:1024M com1=115200,8n1 console=com1,vga
module ${xenvmlinuz} xencons=hvc console=hvc0 console=tty0
module /install.img

boot || goto failed
```

The result of this configuration yields this:
![Screen Shot 2022-07-01 at 6 31 46 PM](https://user-images.githubusercontent.com/12576086/177061689-2e0b7ddf-1837-478c-ad09-d31031bbbc69.png)

Does anyone have any insight of what is needed to get this installer bootable by iPXE? The Citrix site's documentation leaves much to be desired. Once I have a solution for Citrix XenServer 8.2 working I would like to get [XCP-ng](https://xcp-ng.org/) working under iPXE as well.

Please let me know if anyone has any insight in this effort.

Thanks,
Austin

**Comment by stappersg** [comment]
On Sun, Jul 03, 2022 at 05:12:20PM -0700, Austin wrote:
> The result of this configuration yields this:
> ![Screen Shot 2022-07-01 at 6 31 46 PM](https://user-images.githubusercontent.com/12576086/177061689-2e0b7ddf-1837-478c-ad09-d31031bbbc69.png)
> 

VFS: Cannot open root device
Please append a correct "root=" boot option

**Reply by caeles87** [reply]
> VFS: Cannot open root device
Please append a correct "root=" boot option

Thank you for the highlight. I agree that seems to be the main problem, but the question remains how to work with the XenServer initrd that it understands. 

I haven't found any good Citrix documentation on what configuration options it supports out of the box. It would be nice if they had detailed [documentation like Centos 8](https://docs.centos.org/en-US/8-docs/advanced-install/assembly_kickstart-and-advanced-boot-options/#installation-source-boot-options_kickstart-and-advanced-boot-options). 

I have the ISO extracted to a NFS share to try to mount it, but so far all attempts were not successful. I don't even know if the initrd has support for NFS.

**Reply by stappersg** [reply]
> I don't even know if the initrd has support for NFS.

As I interpreter the error messages, are they **not** coming from `initrd`.

**Reply by caeles87** [reply]
Noted. Thanks

**Comment by NiKiZe** [comment]
You have arguments on the initrd (module) line move those arguments to the kernel line.

**Reply by caeles87** [reply]
Thank you for providing the suggestion to attempt.

> You have arguments on the initrd (module) line move those arguments to the kernel line.
In testing this theory I utilized this configuration: 
```
kernel ${xenkernel} dom0_max_vcpus=2  dom0_mem=1024M,max:1024M com1=115200,8n1 console=com1,vga xencons=hvc console=hvc0 console=tty0
module ${xenvmlinuz}
module /install.img
```
This led to pretty much the same output as you can see below:
![Screen Shot 2022-07-05 at 3 58 20 PM](https://user-images.githubusercontent.com/12576086/177417005-eff6ef02-9f85-4930-bd20-99b3ed41d4a7.png)
I have put the original configuration together after reviewing the existing [iPXE XenServer documentation](https://ipxe.org/appnote/xenserver) and combining that with the baked-in isolinux configuration. The baked-in isolinux configuration can be found at `boot/isolinux/isolinux.cfg` in the ISO and can be seen below:
![Screen Shot 2022-07-05 at 4 54 35 PM](https://user-images.githubusercontent.com/12576086/177424481-ae515597-ec47-432c-a715-33c726616005.png)
